kotest-framework/kotest-framework-api/src/jvmTest/kotlin/io/kotest/framework/concurrency/EventuallySpec.kt

```diff
@@ -0,0 +1,368 @@
+
+import io.kotest.assertions.all
+import io.kotest.assertions.assertSoftly
+import io.kotest.assertions.fail
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.assertions.withClue
+import io.kotest.common.ExperimentalKotest
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.ints.shouldBeLessThan
+import io.kotest.matchers.longs.shouldBeGreaterThan
+import io.kotest.matchers.longs.shouldBeGreaterThanOrEqual
+import io.kotest.matchers.nulls.shouldBeNull
+import io.kotest.matchers.nulls.shouldNotBeNull
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldContain
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.async
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.withContext
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.time.Duration
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.reflect.full.createInstance
+import kotlin.system.measureTimeMillis
+
+private fun Int.seconds(): Millis = Duration.ofSeconds(this.toLong()).toMillis()
+private fun Int.milliseconds(): Millis = this.toLong()
+
+@OptIn(ExperimentalKotest::class)
+class EventuallySpec : FunSpec({
+   test("eventually pass working tests") {
+      eventually(5.seconds()) {
+         System.currentTimeMillis()
+      }
+   }
+
+   test("eventually passes tests that completed within the time allowed") {
+      val end = System.currentTimeMillis() + 250L
+      eventually(5.seconds()) {
+         if (System.currentTimeMillis() < end)
+            1 shouldBe 2
+      }
+   }
+
+   test("eventually fails tests that do not complete within the time allowed") {
+      shouldThrow<RuntimeException> {
+         eventually(150L) {
+            throw RuntimeException("foo")
+         }
+      }
+   }
+
+   test("eventually returns the result computed inside") {
+      val result = eventually(2.seconds()) {
+         1
+      }
+      result shouldBe 1
+   }
+
+   test("eventually passes tests that completed within the time allowed, AssertionError") {
+      val end = System.currentTimeMillis() + 250
+      eventually(5.seconds()) {
+         if (System.currentTimeMillis() < end)
+            assert(false)
+      }
+   }
+
+   test("eventually fails tests that throw unexpected exception type") {
+      shouldThrow<NullPointerException> {
+         eventually(2.seconds()).allowExceptions(IOException::class) {
+            (null as String?)!!.length
+         }
+      }
+   }
+
+   test("eventually passes tests that throws FileNotFoundException for some time") {
+      val end = System.currentTimeMillis() + 250
+      eventually(5.seconds()).allowExceptions(FileNotFoundException::class, listener = { true }) {
+         if (System.currentTimeMillis() < end)
+            throw FileNotFoundException("foo")
+      }
+   }
+
```

Comment:
Do we have a test that an exception we dont want immediately fails the block ?

Reply:
```
test("eventually fails tests throw unexpected exception type")
```

this one was existing but doesn't check how many attempts are made since it doesn't get to the part that logs the state of the eventually before finishing up